src/realm/backup_restore.cpp

```diff
@@ -0,0 +1,172 @@
+        if (path[i] == '.')
+            return path.substr(0, i + 1);
+    }
+    // if not on normal "prefix.suffix" form add "."
+    return path + ".";
+}
+
+BackupHandler::BackupHandler(const std::string& path)
+{
+    m_path = path;
+    m_prefix = get_prefix_from_path(path);
+}
+
+bool BackupHandler::must_restore_from_backup(int current_file_format_version)
+{
+    if (current_file_format_version == 0)
+        return false;
+    for (auto i : accepted_versions) {
+        if (i == current_file_format_version)
+            return false;
+        if (backup_exists(m_prefix, i))
+            return true;
+    }
+    return false;
+}
+
+bool BackupHandler::is_accepted_file_format(int version)
+{
+    for (auto i : accepted_versions) {
+        if (i == version)
+            return true;
+    }
+    return false;
+}
+
+void BackupHandler::restore_from_backup()
+{
+    for (auto i : accepted_versions) {
+        if (backup_exists(m_prefix, i)) {
+            auto backup_nm = backup_name(m_prefix, i);
+            // std::cout << "Restoring from:    " << backup_nm << std::endl;
+            util::File::move(backup_nm, m_path);
+        }
+    }
+}
+
+void BackupHandler::cleanup_backups()
+{
+    auto now = time(nullptr);
+    for (auto i : delete_versions) {
+        try {
+            if (backup_exists(m_prefix, i.first)) {
+                std::string fn = backup_name(m_prefix, i.first);
+                // Assuming time_t is in seconds (should be on posix, but...)
+                auto last_modified = util::File::last_write_time(fn);
+                double diff = difftime(now, last_modified);
+                if (diff > i.second) {
+                    // std::cout << "Removing backup:   " << fn << "  - age: " << diff << std::endl;
+                    util::File::remove(fn);
+                }
+            }
+        }
+        catch (...) // ignore any problems, just leave the files
+        {
+        }
+    }
+}
+
+void BackupHandler::backup_realm_if_needed(int current_file_format_version, int target_file_format_version)
+{
+    if (current_file_format_version == 0)
+        return;
+    // FIXME ^^ Is this correct??
+    if (current_file_format_version >= target_file_format_version)
+        return;
+    std::string backup_nm = backup_name(m_prefix, current_file_format_version);
+    if (util::File::exists(backup_nm)) {
+        // std::cout << "Backup file already exists: " << backup_nm << std::endl;
+        return;
+    }
+    // std::cout << "Creating backup:   " << backup_nm << std::endl;
+    std::string part_name = backup_nm + ".part";
+    // FIXME: Consider if it's better to do a writeToFile a la compact?
+    // Silence any errors during the backup process, but should one occur
+    // remove any backup files, since they cannot be trusted.
+    try {
+        util::File::copy(m_path, part_name);
+        util::File::move(part_name, backup_nm);
+    }
+    catch (...) {
+        util::File::try_remove(part_name);
```

Comment:
You should also wrap these in try/catch if you want to ignore all errors, since they can throw in some cases (recently ran into this on windows).